base/event.jl

```diff
@@ -343,3 +343,47 @@ function Timer(cb::Function, timeout::Real, repeat::Real=0.0)
     yieldto(waiter)
     return t
 end
+
+type TimeoutException <: Exception
+    msg::AbstractString
+
+    TimeoutException() = new("")
+    TimeoutException(msg) = new(msg)
+end
+
+"""
+    @set_timeout t ex [msg]
+
+Set a timeout of `t` seconds on the execution time of expression `ex`.
+Throws `TimeoutException` with optional error message `msg`
+if the timeout is triggered.
+"""
+macro set_timeout(timeout_s, ex, msgs...)
+    t_ = eval(timeout_s)
+
+    msg = isempty(msgs) ? ex : msgs[1]
```

Comment:
Cool. I'll take a look. This was based on `macro assert(ex, msgs...)` defined in `error.jl`.